refactor: create a new earn input screen that works with evm tokens

nickewansmith opened the pull request:
## **Description**
This PR refactors the Stake Input component to make it compatible with evm tokens that are not ETH.

We approach this task by starting a fresh Earn directory where we will slowly convert components, utils, hooks etc. to be compatible with all evm tokens. The first flow to tackle is the Stake Input flow. This was moved to the Earn directory and is now Earn Input flow.

Several other files have been moved to the Earn folder and possibly renamed if appropriate. If a file or directory has been moved the counterpart has been removed from the Stake directory and it is now the source of truth. The Earn directory files handle logic whether the `isStableCoinLendingFeatureEnabled`  feature flag is true or false.

The first commit moves some directories over with no logical changes, only path changes if any
The second commit moves some other components, hooks and views and updates logic generally, fixing existing unit tests
The third commit adds unit tests

**This PR does not fully complete the Earn Deposit flow**, but aims to make the component compatible with all evm tokens to prepare for completion in another ticket. While making this component compatible, there was a considerable amount of headway made for the Earn Withdraw flow as well as they use the same input hook. The Earn Withdraw flow is also not fully complete but is prepared for completion in another ticket.
## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/STAKE-949

## **Manual testing steps**

- With stable coin lending flag off, everything should work as it was working before
- With stable coin lending flag off, user should be able to deposit, withdraw and claim 
- With stable coin lending flag off, user should get appropriate errors for not enough eth for gas or not enough for balance
- With stable coin lending flag off, there should be no leaking of stable coin features

- With stable coin lending flag on, user should be able to begin a deposit with any of the tokens in the earn selection list
- With stable coin lending flag on, user should get appropriate errors for not enough token for balance or not enough ETH for gas
  - Note that currently gas estimation for stablecoin lending is not implemented, however, a user should not be able to deposit more than they have. There is also a check for gas that is implemented that allows user to spend max balance of a token that is not ETH as long of they have enough ETH in gas. We will plug in the proper gas estimation in another ticket.
- With stable coin lending flag on, user should be able to begin a withdrawal like normal for ETH as there is no Earn ui yet. Note that if there was evm tokens are now supported

## **Screenshots/Recordings**

https://github.com/user-attachments/assets/d4c3d008-9ada-420d-bd4a-849ed45a547b

### **After**

https://github.com/user-attachments/assets/ef1a5d62-41fe-457c-bc34-ef14c6ee9bf3

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

nickewansmith committed:
refactor: move existing earn specific components to Earn folder

- Moves EarnTokenList, EarnTokenListItem, EarnTokenSelector components to the Earn folder with no logical changes

nickewansmith committed:
refactor: update stake input screen to be compatible with erc20 tokens

- Stake/StakeInputView -> Earn/EarnInputView
- Stake/UnstakeInputView -> Earn/EarnWithdrawInputView
- Move newly compatible files related to Earn input from the Stake folder to the Earn folder and rename
- Move newly compatible files related to Earn Withdraw input from the Stake folder to the Earn folder and possibly rename (EarnWithdrawInputView uses some of the same hooks as EarnInputView and so has also been converted to be erc20 token compatible although it is not fully complete as a feature)
- Removes any refactored Earn folder file from the related Staking folder so there is one source of truth
- Gate the new Earn input functionality behind the stable coin lending feature flag
- Updates existing unit tests

nickewansmith committed:
refactor: add unit tests, clean up

Matt561 commented on app/components/UI/Earn/hooks/useEarnTokenDetails.ts:
```diff
@@ -56,36 +71,49 @@ export const useEarnTokenDetails = () => {
+          tokenBalanceMinimalUnit.toString(),
+        ).toString(),
+        // formatted balance of the asset
+        balanceFormatted: balanceValueFormatted,
+        // fiat balance of the asset
         balanceFiat,
```
Could we add examples for `balanceFormatted` and `balanceFiat`? Maybe it's just me, but I find it difficult sometimes to know what to expect and end up adding console logs to see.

For example, at a glance does `balanceFiat` have a currency symbol in it or is it just the string representation of `balanceFiatNumber`?

nickewansmith replied:
Sure I can add more info to these comments.

Matt561 commented on app/components/UI/Earn/hooks/useEarnTokenDetails.ts:
```diff
@@ -56,36 +71,49 @@ export const useEarnTokenDetails = () => {
+      if (isStablecoinLendingFeatureEnabled()) {
+        apr = MOCK_APR_VALUES[token.symbol] || apr;
+        const rewardRateDecimal = new BigNumber(apr).dividedBy(100).toNumber();
+        const tokenFiatBalanceFloat = parseFloatSafe(token.balanceFiat);
```
Now that we have `assetBalanceFiatNumber` can we remove `tokenFiatBalanceFloat` and `parseFloatSafe`? Are these values the same?

nickewansmith replied:
Yes they are the same, I can remove the usage of `parseFloatSafe` here

Matt561 commented on app/components/UI/Earn/hooks/useInput.ts:
```diff
@@ -0,0 +1,214 @@
+  const [isFiat, setIsFiat] = useState<boolean>(false);
+  // the current selected currency
+  const currentCurrency = useSelector(selectCurrentCurrency);
+  // the current selected chain id
```
Are these comments being auto-added by cursor? Do we want to keep these?

My only concern is that if the code changes and the comments aren't updated they could become misleading or confusing.

nickewansmith replied:
No, ha not Cursor. I made these comments to be clearer about what the variables were. If the code changes, the comments ideally should be updated or removed. I'm fine with removing these comments if the variables are understandable enough.

nickewansmith replied:
Considering you asked for more comment clarification in the useEarnTokenDetails hook, I left these but provided more information.

Matt561 replied:
Okay, no worries at all.

Matt561 commented on app/components/UI/Stake/testUtils/index.ts:
```diff
@@ -26,7 +26,7 @@ export const createMockToken = (options: CreateMockTokenOptions) => {
     chainId,
     name,
     symbol,
-    address = '0xabc',
```
Is this change accidental?

nickewansmith replied:
No it was needed for certain unit tests to pass as there is some check for `safeChecksumAddress` that was making the address matching logic fail in a flow. This is possibly related to it not being a real address, which it still isn't but it does pass that check.

Matt561 commented on app/components/UI/Earn/components/MaxInputModal/index.tsx:
```diff
@@ -48,7 +50,11 @@ const MaxInputModal = () => {
         </BottomSheetHeader>
         <View style={styles.textContainer}>
           <Text variant={TextVariant.BodyMD}>
```
Minor: I noticed that the text is left-aligned when testing locally. Would we want to center-align this instead?

nickewansmith replied:
I'm not opposed to this but would defer to design unless it's a standard in our modals, I can check with Roy as I'm not sure.

nickewansmith replied:
I talked to Roy and his suggestion is to leave for now though he likes the idea as other modals are generally left aligned.

Matt561 replied:
Sounds good

Matt561 commented on app/util/test/renderWithProvider.tsx:
```diff
@@ -101,5 +101,8 @@ export function renderHookWithProvider<Result, Props>(
 
-  return renderHook(hook, { wrapper: Providers } as RenderHookOptions<Props>);
+  return {
+    ...renderHook(hook, { wrapper: Providers } as RenderHookOptions<Props>),
+    store,
```
Curious to know why we added this 👀

nickewansmith replied:
I left this as it might be helpful in the future. I attempted to use it in a test but that path didn't work out too well. This is also exported from `renderWithProvider`

Matt561 reviewed (commented):
Tests well locally. Left a couple of comments and questions but otherwise this looks good!

nickewansmith commented on app/components/UI/Earn/Views/EarnInputView/EarnInputView.tsx:
```diff
@@ -12,46 +16,71 @@ import Button, {
+  const contractExchangeRates = useSelector((state: RootState) =>
+    selectContractExchangeRatesByChainId(state, token.chainId as Hex),
+  );
+  const exchangeRate = contractExchangeRates?.[token.address as Hex]?.price;
```
The `useEarnInputHandlers`  hook defaults to 1 for `exchangeRate`, but I will move it to default to 1 from here in case some other logic uses it further down the line, that would be better.

If the exchange rate is undefined, the pattern I've seen is to default to 1 or 0. I'm defaulting to 1 as there won't be a rate for ETH. 

Thinking of other tokens though, if the `exchangeRate` was undefined then that number might be off if the rate defaulted to 1. 

Instead, I am going to check if `token.isETH` and then set `exchangeRate` to 1 else if an erc20 token set to the rate found in state or 0. This way the rate will be zeroed out for tokens without a rate. 

As far as ui updates for the case, maybe we should explore that separately.

Matt561 replied:
Sounds good!

nickewansmith commented on app/components/UI/Earn/hooks/useEarnInput.ts:
```diff
@@ -0,0 +1,219 @@
+    exchangeRate,
+  });
+
+  const { estimatedGasFeeWei, isLoadingStakingGasFee, isStakingGasFeeError } =
+    useStakingGasFee(balanceWei.toString());
```
Yes, I have created a few TODOs around the specific areas where we will need to address the gas fee for earn tokens and some other issues that may need to be addressed when we connect the contracts to the flow. I like the idea.

The gas hook was intentionally left alone, but once we are at the point of implementing the flow then yes that hook will need to be refactored in some way. I feel there are still some unknowns about the gas fee on my end, so didn't want to move forward with any assumptions about the gas logic for Earn just yet.

Matt561 replied:
Yeah, that's a good call and makes sense to me.

nickewansmith committed:
refactor: adds latest dev ready figma file ui layer

nickewansmith committed:
chore(feedback): address comments, other misc fixes

- updates useEarnTokenDetails such that the formattedEstimatedAnnualRewards change currency symbol along with the selected app settings currency
- updates useEarnTokenDetails apy logic to use the fiat number available instead of the safeParseFloat util
- addresses various comments
- updates unit tests where needed

nickewansmith commented:
> Tests well locally. Left a couple of comments and questions but otherwise this looks good!

@Matt561 Addressed changes requested and added the latest dev ready figma file ui. The new ui is only enabled when the stable coin lending flag is on.

nickewansmith committed:
chore(feedback): ui and unit test updates

Matt561 commented on app/components/UI/Earn/Views/EarnWithdrawInputView/EarnWithdrawInputView.tsx:
```diff
@@ -12,85 +15,161 @@ import Button, {
+    hasBackButton: true,
+    hasIconButton: true,
+    // TODO: STAKE-903
+    // handleIconPress: ???,
```
I'll make sure the graph is added as part of [STAKE-967](https://consensyssoftware.atlassian.net/browse/STAKE-967) 👍

nickewansmith committed:
Merge branch 'main' into STAKE-949-fe-create-a-new-earn-input-screen-that-works-with-multiple-tokens

chrisleewilcox reviewed (approved):
LGTM
